src/util/text-utils.cc

```diff
@@ -165,90 +167,89 @@ inline bool starts_with(const std::string &in, const std::string &prefix) {
+    // we'll keep just lowercase values.
+    inf_nan_map["inf"] = std::numeric_limits<T>::infinity();
+    inf_nan_map["infinity"] = std::numeric_limits<T>::infinity();
+    inf_nan_map["nan"] = std::numeric_limits<T>::quiet_NaN();
+
+    std::string c;
+    in_.clear();
+    // If the stream is broken even before trying
+    // to read from it, it's pointless to try.
+    if (!(in_ >> c)) return *this;
+
+    // transform c to lowercase.
+    std::transform(c.begin(), c.end(), c.begin(), ::tolower);
+
+    if (inf_nan_map.find(c) != inf_nan_map.end()) {
+      *x = inf_nan_map[c];
+      if (neg) *x = - *x;
+    } else {
+      in_.setstate(std::ios_base::failbit);
+    }
 
-template<typename T>
-bool convert_special_number(const std::string &str, T *out) {
-  if (stricmp(str, "infinity") || stricmp(str, "inf") ||
-      starts_with(str, "1.#INF")) {
-    *out = std::numeric_limits<T>::infinity();
-    return true;
-  } else if (stricmp(str, "-infinity") || stricmp(str, "-inf") ||
-             starts_with(str, "-1.#INF")) {
-    *out = -std::numeric_limits<T>::infinity();
-    return true;
-  } else if (is_nan_text(str, "nan") || starts_with(str, "1.#QNAN")) {
-    *out = std::numeric_limits<T>::quiet_NaN();
-    return true;
-  } else if (is_nan_text(str, "-nan") || starts_with(str, "-1.#QNAN")) {
-    *out = -std::numeric_limits<T>::quiet_NaN();
-    return true;
+    return *this;
   }
-  return false;
-}
+};
 
+
+template <typename T>
 bool ConvertStringToReal(const std::string &str,
-                         double *out) {
-  const char *this_str = str.c_str();
-  char *end = NULL;
-  errno = 0;
-
-#if defined(_MSC_VER)
-  // TODO: check if the new MSVC already supports it
-  // depending on claims of the C++11 support, it should have
-  if (convert_special_number(str, out))
-    return true;
-#endif  // defined(_MSC_VER)
+                         T *out) {
+  std::istringstream iss(str);
 
-  double d = KALDI_STRTOD(this_str, &end);
-  if (end != this_str)
-    while (isspace(*end)) end++;
-  if (end == this_str || *end != '\0' || errno != 0)
-    return false;
-  *out = d;
-  return true;
-}
+  NumberIstream<T> i(iss);
 
-bool ConvertStringToReal(const std::string &str,
-                         float *out) {
-  const char *this_str = str.c_str();
-  char *end = NULL;
-  errno = 0;
-
-#ifdef _MSC_VER
-  // TODO: check if the new MSVC already supports it
-  // depending on claims of the C++11 support, it should have
-  if (convert_special_number(str, out))
-    return true;
-#endif  // _MSC_VER
+  i >> *out;
 
-  float f = KALDI_STRTOF(this_str, &end);
-  if (end != this_str)
-    while (isspace(*end)) end++;
-  if (end == this_str || *end != '\0' || errno != 0)
+  if (iss.fail()) {
+    // Number conversion failed.
     return false;
-  *out = f;
+  }
+
+  // if something remains in the istringstream,
+  // we'll check if it is #INF or #QNAN (to deal with
+  // MSVC stuffs), or if it is some garbage text.
+  if (iss.tellg() != -1) {
```

Comment:
I don't understand why you are putting the different types of infinity (the MSVC ones and the regular ones) in different places.  It makes the code quite ugly.  And also I think you should be skipping over other kinds of space character, not just ' '.

Also, it seems to me that the way this code is written, it would falsely accept strings that consisted of some kind of garbage followed by `#INF`: for instance, `hello #INF`.